File: backend/app/schemas/career_copilot.py
from uuid import UUID

from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.cv_optimizer import JobProvider


CareerCopilotActionType = Literal[
    "open_cv_optimizer",
    "open_jobs",
    "open_roadmap",
    "open_profile",
    "open_upload_cv",
    "open_history",
]


class CareerCopilotRequest(BaseModel):
    analysis_id: UUID
    message: str = Field(min_length=1, max_length=2000)
    job_external_id: str | None = None
    provider: JobProvider | None = None

    @field_validator("message")
    @classmethod
    def validate_message(cls, value: str) -> str:
        normalized_value = value.strip()

        if not normalized_value:
            raise ValueError("Message is required.")

        return normalized_value

    @field_validator("job_external_id")
    @classmethod
    def normalize_optional_job_external_id(cls, value: str | None) -> str | None:
        if value is None:
            return None

        normalized_value = value.strip()
        return normalized_value or None


class CareerCopilotSuggestedAction(BaseModel):
    type: CareerCopilotActionType
    label: str
    target: str


class CareerCopilotToolResult(BaseModel):
    type: Literal["cv_optimization"]
    status: Literal["completed", "requires_input", "failed"]
    data: dict[str, Any] | None = None


class CareerCopilotResponse(BaseModel):
    reply: str
    suggested_action: CareerCopilotSuggestedAction | None = None
    tool_result: CareerCopilotToolResult | None = None

File: src/types/careerCopilot.ts
import type { CVOptimizeResponse, CVOptimizerProvider } from './cvOptimizer';

export type CareerCopilotRequest = {
  analysis_id: string;
  message: string;
  job_external_id?: string | null;
  provider?: CVOptimizerProvider | null;
};

export type CareerCopilotActionType =
  | 'open_cv_optimizer'
  | 'open_jobs'
  | 'open_roadmap'
  | 'open_profile'
  | 'open_upload_cv'
  | 'open_history';

export type CareerCopilotSuggestedAction = {
  type: CareerCopilotActionType;
  label: string;
  target: string;
};

export type CareerCopilotResponse = {
  reply: string;
  suggested_action: CareerCopilotSuggestedAction | null;
  tool_result: CareerCopilotToolResult | null;
};

export type CareerCopilotCVOptimizationToolData = {
  current_match: number;
  estimated_match: number;
  changes: string[];
  major_changes?: string[];
  optimization_summary?: string;
  before_professional_summary?: string;
  optimized_professional_summary?: string;
  optimized_skills?: string[];
  explanation?: string;
  optimized_cv: CVOptimizeResponse['optimized_cv'];
};

export type CareerCopilotToolResult = {
  type: 'cv_optimization';
  status: 'completed' | 'requires_input' | 'failed';
  data: CareerCopilotCVOptimizationToolData | null;
};

export type CareerCopilotMessage = {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  createdAt: string;
  suggestedAction?: CareerCopilotSuggestedAction | null;
  toolResult?: CareerCopilotToolResult | null;
};
